### ISSUE TITLE
`--show-source` code frames should respect Jupyter cells

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ruff#5395]
I have two cells, one containing two imports, and another containing a function definition.

When I run `cargo run -p ruff_cli -- check Untitled.ipynb -n --watch --show-source`, I get:

```
Untitled.ipynb:cell 1:2:8: F401 [*] `sys` imported but unused
  |
1 | import os
2 | import sys
  |        ^^^ F401
3 | def f(x):
4 |     y = x * x
  |
  = help: Remove unused import: `sys`

Untitled.ipynb:cell 2:2:5: F841 [*] Local variable `y` is assigned to but never used
  |
2 | import sys
3 | def f(x):
4 |     y = x * x
  |     ^ F841
5 |     print(x * x)
6 | f(2)
  |
  = help: Remove assignment to unused variable `y`
```

I think the code frames need to be reindexed. Notice that in the second error, the location is reported as `cell 2:2:5`, but in the code frame, the violation is on line 4.

Similarly, we should omit any contents from adjoining cells when displaying the code frame (in my opinion) -- so the second diagnostic shouldn't show the `import sys`.

**Comment by charliermarsh** [comment]
\cc @dhruvmanila - Wdyt?

**Comment by dhruvmanila** [comment]
Ah yes, this is true. I'll need to look into how source is being shown currently. Most likely the cell separation concept could be common for `--show-source` and `--diff` option.
